[Docs] ES UI form library (#78654)
elastic/kibana: src/plugins/es_ui_shared/docs/docs/form_lib/core/form_hook.md

This object has the following properties and handlers

## Properties

### isSubmitted

**Type:** `boolean`

Flag that indicates if the form has been submitted at least once. It is set to `true` when we call `submit()`. ([see below](#submit)).  
**Note:** If you have a dynamic form where fields are removed and added, the `isSubmitted` is set to `false` whenever a new field is added, as in such case the user has a new form in front of him.

### isSubmitting

**Type:** `boolean`

Flag that indicates if the form is being submitted. When we submit the form, if you have provided an [`onSubmit()` handler](use_form_hook.md#onsubmitdata-isvalid) in the config, it might take some time to resolve (e.g. an HTTP request being made). This flag will be set to `true` until the Promise resolves.

### isValid

**Type:** `boolean | undefined`

Flag that indicates if the form is valid. It can have three values:
* `true`
* `false`
* `undefined`

When the form first renders, its validity is neither `true` nor `false`. It is `undefined`, we don't know its validity. It could be valid if none of the fields are required or invalid if some field is required.

> Each time a field value changes, it is validated. When **all** fields have changed (are dirty), then only the `isValid` is either `true` or `false`, as at this stage we know the form validity. Of course we will probably need to know the validity of the form without updating each field one by one. There are two ways of doing that:

Reviewer: 👍 I will add a note below next to the `validate()` bullet

      <UseField path="firstName" config={{ validations: [{ validator }] }} />
      <UseField path="lastName" />

      <button
        onClick={onClickSubmit}
        disabled={isSubmitting || hasErrors}
      >
        {isSubmitting ? 'Sending...' : 'Submit'}
      </button>
      {hasErrors && <div>Form is invalid.</div>}
    </Form>
  );
};
```

### id

**Type:** `string`

The form id. If none was provided, "default" will be returned.

## Handlers

### submit()

**Returns:** `Promise<{ data: T | {}, isValid: boolean }>`

This handlers submits the form and returns its data and validity. If the form is not valid, the data will be `null` as only valid data is passed through the `serializer(s)` before being returned.

```js
const { data, isValid } = await form.submit();
```

### validate()

**Returns:** `Promise<boolean>`

Use this handler to get the validity of the form.

```js
const isFormValid = await form.validate();
```

### getFields()

**Returns:** `{ [path: string]: FieldHook }`

Access any field on the form.

```js
const { name: nameField }  = form.getFields();
```

### getFormData()

**Arguments:** `options?: { unflatten?: boolean }`  
**Returns:** `T | R`

> Return the form data. It accepts an optional options with an `unflatten` parameter (defaults to `true`). If you are only interested in the raw form data, pass `unflatten: false` to the handler ([read more about the "Raw" data state here](in_out_raw_state.md#raw-data-state)).

Reviewer 2: ```suggestion
Return the form data. Accepts an optional `options` with an `unflatten` parameter (defaults to `true`). If you are only interested in the raw form data, pass `unflatten: false` to the handler ([read more about the "Raw" data state here](in_out_raw_state.md#raw-data-state)).
```

Reviewer 2: I guess this is going away after: https://github.com/elastic/kibana/issues/78589 ?

Might be good to make a not of that as part of the work to be done along with that PR change (i.e., that the docs need to be updated)

Reviewer: Good point, I will add a note for it in the PR 👍

```js
const formData  = form.getFormData();

const rawFormData  = form.getFormData({ unflatten: false });
```

### getErrors()

**Returns:** `string[]`

Returns an array with of all errors in the form.

```js
const errors  = form.getErrors();
```

### reset()

**Arguments:** `options?: { resetValues?: boolean; defaultValue?: any }` 

Resets the form to its initial state. It accetps an optional object of configuration:

> - `resetValues` (default: `true`). Flag to indicate if we want to not only reset the form state (`isValid`, `isSubmitted`...) but also the field values. If set to `true` will put back the default value passed to the form or declared on the field config (in that order).

Reviewer 2: ```suggestion
- `resetValues` (default: `true`). Flag to indicate if we want to not only reset the form state (`isValid`, `isSubmitted`...) but also the field values. If set to `true` all form values will be reset to their default value. Remember: default field values can be set in [different ways](./link-to-place-where-we-talk-about-it).
```

> - `defaultValue`. In some cases you might not want to reset the form to the default value initiallly provided to the form (probably because it is data that came from the server and you want a clean form). In this case you can provide a new `defaultValue` object when resetting.

Reviewer 2: Might also be worth adding a word of caution here that using this to imperatively set form state can lead to users form data being lost and that this is a kind of escape hatch.

Reviewer 2: Maybe we can even call it something like `dangerouslySetDefaultValues` if we feel strongly enough 😄

Reviewer: I am not sure I follow :blush: If the consumer specifically calls `form.reset()` in his code, isn't it implicit that the values will be all changed (and lost)?
